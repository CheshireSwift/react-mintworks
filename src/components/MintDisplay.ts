import { Fragment, CSSProperties } from 'react';
import { h, span } from 'react-hyperscript-helpers';
import * as R from 'ramda';
import emoji from 'node-emoji';

const Mint = (props: { bigMint: boolean; first: boolean }) =>
  span(
    { style: { marginLeft: props.first ? 0 : -16 } },
    emoji.get(props.bigMint ? 'red_circle' : 'white_circle')
  );

const firstMint = (bigMint: boolean, count: number) =>
  !!count && h(Mint, { bigMint, first: true });

const repeatMints = (bigMint: boolean) =>
  R.pipe(
    R.max<number>(0),
    R.repeat(h(Mint, { bigMint, first: false }))
  );

export const MintDisplay = ({
  count,
  style
}: {
  count: number;
  style?: CSSProperties;
}) => {
  const hasBigMints = count >= 5;
  return span({ style }, [
    firstMint(hasBigMints, count),
    ...repeatMints(true)(Math.floor(count / 5) - 1),
    ...repeatMints(false)((hasBigMints ? count : count - 1) % 5),
    count
  ]);
};

export default MintDisplay;
