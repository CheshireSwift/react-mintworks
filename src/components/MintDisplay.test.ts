import { render } from 'enzyme';
import { h } from 'react-hyperscript-helpers';
import emoji from 'node-emoji';

import MintDisplay from './MintDisplay';

const mintIcon = emoji.get('white_circle');
const bigMintIcon = emoji.get('red_circle');

describe('Mint display', () => {
  const mintDisplayText = render(h(MintDisplay, { count: 3 })).text();

  it('matches the snapshot', () => {
    expect(
      render(h(MintDisplay, { count: 16, style: { fontSize: 'larger' } }))
    ).toMatchSnapshot();
  });

  it('shows the mint count visually + numerically', () => {
    expect(mintDisplayText).toEqual(mintIcon + mintIcon + mintIcon + '3');
  });

  it('uses red mints for high mint counts', () => {
    const bigMintDisplayText = render(h(MintDisplay, { count: 17 })).text();

    expect(bigMintDisplayText).toEqual(
      bigMintIcon + bigMintIcon + bigMintIcon + mintIcon + mintIcon + '17'
    );
  });
});
