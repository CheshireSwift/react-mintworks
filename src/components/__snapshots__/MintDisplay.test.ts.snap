// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mint display matches the snapshot 1`] = `
<span
  style="font-size:larger"
>
  <span
    style="margin-left:0"
  >
    ðŸ”´
  </span>
  <span
    style="margin-left:-16px"
  >
    ðŸ”´
  </span>
  <span
    style="margin-left:-16px"
  >
    ðŸ”´
  </span>
  <span
    style="margin-left:-16px"
  >
    âšª
  </span>
  16
</span>
`;
